File: news_server/modules/source/infrastructure/getters/zhihu.ts
/**
 * 知乎新闻源获取器
 * 从 newsnow/server/sources/zhihu.ts 迁移
 */
import type { SourceGetter } from '../../domain/getter/getter'
import { NewsItem } from '../../domain/newsitem/entity'

interface ZhihuResponse {
  data: {
    type: 'hot_list_feed'
    style_type: '1'
    feed_specific: {
      answer_count: number
    }
    target: {
      title_area: {
        text: string
      }
      excerpt_area: {
        text: string
      }
      image_area: {
        url: string
      }
      metrics_area: {
        text: string
        font_color: string
        background: string
        weight: string
      }
      label_area: {
        type: 'trend'
        trend: number
        night_color: string
        normal_color: string
      }
      link: {
        url: string
      }
    }
  }[]
}

export const zhihuGetter: SourceGetter = async () => {
  const url = 'https://www.zhihu.com/api/v3/feed/topstory/hot-list-web?limit=20&desktop=true'
  const response = await fetch(url)
  
  if (!response.ok) {
    throw new Error(`Failed to fetch Zhihu feed: ${response.statusText}`)
  }

  const res: ZhihuResponse = await response.json()

  return res.data.map((item) => {
    const urlMatch = item.target.link.url.match(/(\d+)$/)
    const id = urlMatch ? urlMatch[1] : item.target.link.url

    return NewsItem.fromEntity({
      id,
      title: item.target.title_area.text,
      url: item.target.link.url,
      extra: {
        info: item.target.metrics_area.text,
        hover: item.target.excerpt_area.text,
      },
    })
  })
}

File: modules/source/infrastructure/getters/zhihu.go
package getters

import (
	"context"
	"encoding/json"
	"regexp"
)

func (r *Registry) zhihu(ctx context.Context) ([]Item, error) {
	body, err := r.get(ctx, "https://www.zhihu.com/api/v3/feed/topstory/hot-list-web?limit=20&desktop=true", nil)
	if err != nil {
		return nil, err
	}
	var res struct {
		Data []struct {
			Target struct {
				TitleArea   struct{ Text string `json:"text"` } `json:"title_area"`
				ExcerptArea struct{ Text string `json:"text"` } `json:"excerpt_area"`
				MetricsArea struct{ Text string `json:"text"` } `json:"metrics_area"`
				Link        struct{ URL string `json:"url"` }  `json:"link"`
			} `json:"target"`
		} `json:"data"`
	}
	if err := json.Unmarshal(body, &res); err != nil {
		return nil, err
	}
	re := regexp.MustCompile(`(\d+)$`)
	out := make([]Item, 0, len(res.Data))
	for _, k := range res.Data {
		id := k.Target.Link.URL
		if m := re.FindString(k.Target.Link.URL); m != "" {
			id = m
		}
		out = append(out, Item{
			ID: id, Title: k.Target.TitleArea.Text, URL: k.Target.Link.URL,
			Extra: map[string]any{"info": k.Target.MetricsArea.Text, "hover": k.Target.ExcerptArea.Text},
		})
	}
	return out, nil
}
